### ISSUE TITLE
Could each CSS names (elements, ids and classes) have specific types?

### ISSUE THREAD

**Issue by cies** [slamdata/purescript-css#41]
I guess it is not possible to implement this idea --if at all possible-- in `purescript-css`, it will need a library of it's own.

What I mean by specific types for CSS names can be seen in this Elm library, [elm-css](https://github.com/rtfeldman/elm-css). It allows type checking of ids and classes between your style definitions and HTML templates. Which to me seems like an amazingly good idea. :)

Possibly a library that implements this already exist for PureScript, but that I'm simply unaware of it.

**Comment by garyb** [comment]
Do you mean introducing `newtype`s to distinguish between them or something? Sorry, I'm not quite sure what you're getting at. If so, sounds good to me!

**Comment by cies** [comment]
Please have a look at the elm-css lib. It petty much explains itself.

I would not know if this can be achieved with only newtypes. I guess not as
you want to distinguish between ids and classes.

**Comment by cies** [comment]
Ok, more time to go indepth now. :)

See here the defintion of some CSS using `elm-css`.

```elm
type CssClasses = NavBar
type CssIds = Page

css =
    (stylesheet << namespace "dreamwriter")
    [ body
        [ overflowX auto
        , minWidth (px 1280)
        ]
    , (#) Page
        [ backgroundColor (rgb 200 128 64)
        , color (hex "CCFFFF")
        , width (pct 100)
        , height (pct 100)
        , boxSizing borderBox
        , padding (px 8)
        , margin zero
        ]
    , (.) NavBar
        [ margin zero
        , padding zero
        , children
            [ li
                [ (display inlineBlock) |> important
                , color primaryAccentColor
                ]
            ]
        ]
    ]
```

Pretty similar to what it looks like with `purescript-css`, but with some actual types for `NavBar` and `Page`. Now these will be available when rendering HTML.

```elm
import Html.CssHelpers
import MyCss

{ id, class, classList } =
    Html.CssHelpers.withNamespace "dreamwriter"

view =
    Html.div []
        [ Html.div [ class [ MyCss.NavBar ] ] [ Html.text "this has the NavBar class" ]
        , Html.div [ id MyCss.Page ] [ Html.text "this has the Page id" ]
        ]
```

This brings typesafety to the rescue for typos in CSS-HTML id/class names, and allows auto-complete in editors to help you out more.

**Comment by garyb** [comment]
Ah! I see now :smile:. I'm sure we could accomodate something like this, yes. I don't really have time to work on it right now, but it seems like a good idea.

**Comment by cies** [comment]
Ok @garyb, I found some time for it and have a little to show you. It is currently all implemented as part of the  test suite, just as a proof-of-concept, mainly for you (and other to shoot at). It likely have to be implemented all differently, or there's an easier way to implemented it, or it can be more well-typed.

Please [have a look](https://github.com/cies/purescript-css/blob/d29242171c4e69ce0cec4bd476f087d15a241e43/test/Main.purs#L81). I commented it with explanation.

**Comment by cies** [comment]
Oh and I really like to change this:

```purescript
data CssId = TheWrapper | TheContent
derive instance genericCssId :: Generic CssId
instance cssPredicateCssId :: CssPredicate CssId where
  toPredicate = Id <<< spineCase <<< typeAsString

data CssClass = ButtonSEO | Active
derive instance genericCssClass :: Generic CssClass
instance cssPredicateCssClass :: CssPredicate CssClass where
toPredicate = Class <<< spineCase <<< typeAsString
```

into something more like (I know this does not work):

```purescript
data CssId = TheWrapper | TheContent
derive instance cssIdPredicates :: CssPredicate CssId

data CssClass = ButtonSEO | Active
derive instance cssClassPredicated :: CssPredicate CssClass
```

but I did not yet figure out how.

**Comment by garyb** [comment]
Yeah, as you say, you can't derive like that, but with the way you've used generics there you could just define two common functions:

``` purescript
toIdPredicate = Class <<< spineCase <<< typeAsString
toClassPredicate = Class <<< spineCase <<< typeAsString
```

And then those would work generically for the `toPredicate` implementation of any class:

``` purescript
data CssId = TheWrapper | TheContent
derive instance genericCssId :: Generic CssId
instance cssPredicateCssId :: CssPredicate CssId where
  toPredicate = toIdPredicate 

data CssClass = ButtonSEO | Active
derive instance genericCssClass :: Generic CssClass
instance cssPredicateCssClass :: CssPredicate CssClass where
  toPredicate = toClassPredicate
```

At least that cuts down on the boilerplate for `toPredicate` implementations.

I wouldn't include `Generic` as a superclass of `CssPredicate` either, as it's not required for the definition, it just makes it easier to implement `toPredicate` - so if you make it optional, and only have the `Generic` constraint on `toIdPredicate` / `toClassPredicate` it will allow people to write their own `toPredicate` implementations if they really need to, without using generic deriving.
